File: src/main/java/io/citytrees/repository/FileContentRepository.java
package io.citytrees.repository;

import io.citytrees.model.CtFileContent;
import org.springframework.data.jdbc.repository.query.Modifying;
import org.springframework.data.jdbc.repository.query.Query;
import org.springframework.data.repository.CrudRepository;

import java.util.Optional;
import java.util.UUID;

public interface FileContentRepository extends CrudRepository<CtFileContent, UUID> {

    @Query("""
        INSERT INTO ct_file_content(id, content, hash)
        VALUES (:id, :content, :hash)
        RETURNING id
        """)
    UUID save(UUID id, byte[] content, String hash);

    @Modifying
    @Query("""
        INSERT INTO ct_file_content(id, content, hash)
        VALUES (:id, :content, :hash)
        ON CONFLICT(hash)
        DO NOTHING;
        """)
    int upsert(UUID id, byte[] content, String hash);

    Optional<CtFileContent> findFirstByHash(String hash);

    void deleteByHash(String hash);
}

File: src/main/java/io/citytrees/repository/FileContentRepository.kt
package io.citytrees.repository

import io.citytrees.constants.TableNames.FILE_CONTENT_TABLE
import io.citytrees.model.CtFileContent
import org.springframework.data.jdbc.repository.query.Modifying
import org.springframework.data.jdbc.repository.query.Query
import org.springframework.data.repository.CrudRepository
import java.util.*

interface FileContentRepository : CrudRepository<CtFileContent, UUID> {

    fun findFirstByHash(hash: String): Optional<CtFileContent>

    fun deleteByHash(hash: String)

    @Modifying
    @Query(
        """
        INSERT INTO $FILE_CONTENT_TABLE(id, content, hash)
        VALUES (:id, :content, :hash)
        ON CONFLICT(hash)
        DO NOTHING
        """
    )
    fun upsert(id: UUID, content: ByteArray, hash: String): Int
}